### PR TITLE
Add SSL/TLS encryption in tcpclient

### PR DESCRIPTION
**Pull request by mbhangui**
This PR adds **SSL/TLS** encryption in tcpclient. A client program dotls has been included which reads stdin, writes to fd _7_ and reads fd _6_ and writes to stdout

This PR also adds client mode operation when prog is omitted. In this case tcpclient passes data from fd _0_ (_stdin_) to the network and data from the network to fd _1_ (_stdout_). This makes it work like **telnet**(1), **ssh**(1) and other programs

**Examples**

1. Connect to port 25
```
$ tcpclient 0 25
220 indimail.org (NO UCE) ESMTP IndiMail 1.238 Wed, 3 Mar 2021 18:08:10 +0530
ehlo
250-indimail.org [::1]
250-PIPELINING
250-8BITMIME
250-SIZE 10000000
250-ETRN
250-STARTTLS
250 HELP
quit
221 indimail.org closing connection
```

2. Connect to port SMTPS port on 465 using IPv6
```
$ tcpclient -n "" ::1 465
220 indimail.org (NO UCE) ESMTP IndiMail 1.238 Wed, 3 Mar 2021 18:09:52 +0530
ehlo
250-indimail.org [::1]
250-AUTH LOGIN PLAIN CRAM-MD5 CRAM-SHA1 CRAM-SHA256 CRAM-SHA512 CRAM-RIPEMD DIGEST-MD5
250-AUTH=LOGIN PLAIN CRAM-MD5 CRAM-SHA1 CRAM-SHA256 CRAM-SHA512 CRAM-RIPEMD DIGEST-MD5
250-PIPELINING
250-8BITMIME
250-SIZE 10000000
250-ETRN
250 HELP
quit
221 indimail.org closing connection
```

3. Provide oppurtunistic TLS for notqmail/netqmail/qmail qmail-smtpd
```
$ tcpserver -HR -v 0 5555 /usr/bin/dotls -n "" -s smtp /var/qmail/bin/qmail-smtpd
tcpserver: status: 0/40 sql: 1
```

4. Run tcpclient in client mode and issue STARTTLS command. This is similar to openssl s_client -starttls -crlf
```
$ tcpclient -HRv -n "" -s smtp 0 5555
tcpclient: connected to ::1 port 5555
220 indimail.org ESMTP
ehlo
250-indimail.org
250-PIPELINING
250 8BITMIME
help
214 notqmail home page: https://notqmail.org
quit
221 indimail.org
```

**Files modified**
tcpclient.c
tcpserver.c
tcpclient.9
tcpserver.9
Makefile
.gitignore
TARGETS

**Files added**
tls.c
tls.h
dotls.c
dotls.1

### CODE REVIEW

File: ucspi-tcp-x/dotls.c [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+#define FATAL         "sslerator: fatal: "
+#define HUGESMTPTEXT  5000
+
+#ifndef	lint
+static char     sccsid[] = "$Id: $";
+#endif
+
+struct substdio ssin, ssto, smtpin, smtpto;
+unsigned long   dtimeout = 60;
+stralloc        certfile = { 0 };
```
Comment (DerDakon) [inline ucspi-tcp-x/dotls.c:38]: global variables will be 0-initialized by default anyway.

File: ucspi-tcp-x/dotls.c [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+			strerr_die2x(111, FATAL, "out of memory");
+	}
+	scan_ulong(str, &code);
+	return code;
+}
+
+unsigned long
+smtpcode(unsigned int *len)
+{
+	unsigned char   ch;
```
Comment (DerDakon) [inline ucspi-tcp-x/dotls.c:150]: If you remove the unsigned here you can remove the later casts.

File: ucspi-tcp-x/dotls.c (hunk 1 above)
Comment (mbhangui) [reply ucspi-tcp-x/dotls.c:38]: This is a bad habit I have of initializing variables :).

File: ucspi-tcp-x/tcpserver.c [hunk 3]
```diff
@@ -1347,10 +1354,10 @@ main(int argc, char **argv, char **envp)
 #endif
 #endif
 #ifdef TLS
-	BIO            *sbio;
 	SSL            *ssl;
-	SSL_CTX        *ctx;
-	char           *controldir;
+	SSL_CTX        *ctx = (SSL_CTX *) NULL;
+	char           *certsdir, *ciphers = (char *) NULL,
```
Comment (DerDakon) [inline ucspi-tcp-x/tcpserver.c:1359]: These casts are not necessary in C.

**Pull request merged**